### PR TITLE
Icon handling

### PR DESCRIPTION
**Pull request by davidnuescheler**
this is a client sided implementation of https://github.com/adobe/helix-pipeline-service/issues/386
to see if it is feasible.

https://icon-handling--helix-project-boilerplate--adobe.hlx.page/

as there are currently no icons in the project there is not much to see here.

### CODE REVIEW

**Comment by davidnuescheler** [comment]
@kptdobe i am sure that the `await` in the `.forEach` loop could be improved, and i am sure that you might have the best solution handy.

**Comment by davidnuescheler** [comment]
i think this is easiest to see in action on here https://nav-refactor--helix-project-boilerplate--adobe.hlx.page/nav-new as this includes an icon in the nav that is used in different places with different CSS context.

File: scripts/scripts.js [hunk 1]
```diff
@@ -120,6 +120,36 @@ export function toClassName(name) {
...
+    span.className = img.className;
+    img.replaceWith(span);
+  });
+}
+
+/**
+ * Replace icons with inline SVG and prefix with codeBasePath.
+ * @param {Element} element
+ */
+export async function decorateIcons(element) {
```
Comment (kptdobe) [inline scripts/scripts.js:139]:
```suggestion
export function decorateIcons(element) {
```

As is, the method is not `async`.

File: blocks/header/header.js [hunk 2]
```diff
@@ -64,6 +64,6 @@ export default async function decorate(block) {
   });
   nav.prepend(hamburger);
   nav.setAttribute('aria-expanded', 'false');
-
+  await decorateIcons(nav);
```
Comment (kptdobe) [inline blocks/header/header.js:67]:
```suggestion
  decorateIcons(nav);
```

As is, the method is not `async`.

**Comment by kptdobe** [comment]
> @kptdobe i am sure that the await in the .forEach loop could be improved, and i am sure that you might have the best solution handy.

It depends. I think it is fine as is because we do not care of the order and the promises will be executed by the browser (not risk to exit before they are all executed). This should just be reflected in the `decorateIcons` method signature: it is not async.
If you really want to make it async, then yes, you need to change the loop to something like (not tested):

```
export async function decorateIcons(element) {
  replaceIcons(element);

  const promises = [];
  element.querySelectorAll('span.icon').forEach((span) => {
    promises.push(new Promise((resolve) => {
      const iconName = span.className.split('icon-')[1];
      const resp = await fetch(`${window.hlx.codeBasePath}/icons/${iconName}.svg`);
      if (resp.status === 200) {
        const svg = await resp.text();
        span.innerHTML = svg;
      }
      resolve();
    });
  });

  await Promise.all(promises);
}
```

File: scripts/scripts.js (hunk 1 above)
Comment (rofe) [reply scripts/scripts.js:139]: What about the `await fetch()`?

**Pull request merged**